File: com.crypto.action/Bitcoin.java
package com.crypto.action;

import java.math.BigInteger;
import java.security.MessageDigest;
import java.security.Security;

import org.bitcoinj.core.Base58;
import org.bouncycastle.jce.provider.BouncyCastleProvider;

import com.crypto.entity.Point;

public class Bitcoin {
	
	public static void main(String[] args) throws Exception {
		
		BigInteger base = new BigInteger("2");
		
		BigInteger mod = base.pow(256)
				.subtract(base.pow(32))
				.subtract(base.pow(9))
				.subtract(base.pow(8))
				.subtract(base.pow(7))
				.subtract(base.pow(6))
				.subtract(base.pow(4))
				.subtract(base.pow(0));
		
		BigInteger order = new BigInteger("FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEBAAEDCE6AF48A03BBFD25E8CD0364141", 16); 
		
		//curve equation: y^2 = x^3 + ax + b -> current curve: y^2 = x^3 + 7
		BigInteger a = new BigInteger("0");
		BigInteger b = new BigInteger("7");
		
		System.out.println("Curve: y^2 = x^3 + "+a+"*x + "+b);
		
		Point basePoint = new Point();
		basePoint.setPointX(new BigInteger("55066263022277343669578718895168534326250603453777594175500187360389116729240"));
		basePoint.setPointY(new BigInteger("32670510020758816978083085130507043184471273380659243275938904335757337482424"));
		
		System.out.println("base point: ("+basePoint.getPointX()+", "+basePoint.getPointY()+")\n");
		
		System.out.println("modulo: "+mod);
		System.out.println("order of group: "+order+"\n");
		
		System.out.println("--------------------------------");
		System.out.println("Bitcoin Address Generator");
		System.out.println("--------------------------------\n");
		
		//BigInteger privateKey = new BigInteger("11253563012059685825953619222107823549092147699031672238385790369351542642469");
		
		//BigInteger privateKey = new BigInteger("18E14A7B6A307F426A94F8114701E7C8E774E7F9A47E2C2035DB29A206321725", 16);
		//BigInteger privateKey = new BigInteger("945FB84E92575908A000D9D4A0B136EB0B69F8B5BCDAF157A3A91B7C83A02B1B", 16);
		BigInteger privateKey = new BigInteger("118CC1421845A67643244180F299147659E55C439B34E9E2380946579F64EF16", 16);
		
		System.out.println("private key (hex): "+privateKey.toString(16)+" ("+privateKey.toString(16).length()*4+" bits)\n");
		
		Point publicKey = EccOverFiniteField.applyDoubleAndAddMethod(basePoint, privateKey, a, b, mod);
		
		System.out.println("public key: ("+publicKey.getPointX()+", "+publicKey.getPointY()+")");
		
		String publicKeyString = "04"+publicKey.getPointX().toString(16)+publicKey.getPointY().toString(16);
		
		BigInteger publicKeyMerged = new BigInteger(publicKeyString, 16);
		
		System.out.println("public key (merged): "+publicKeyString+" ("+publicKeyString.length()*4+" bits)\n");
		
		System.out.println("--------------------------------\n");
		
		MessageDigest md = MessageDigest.getInstance("SHA-256");
		md.update(publicKeyMerged.toByteArray());
		byte[] sha256HashByte = md.digest();
		
		BigInteger hash = new BigInteger(sha256HashByte).abs();
		
		System.out.println("sha-256 applied hash: "+hash.toString(16)+" ("+hash.toString(16).length()*4+" bits)");
		
		//-------------------------------------------------------
		
		Security.addProvider(new BouncyCastleProvider());
		
		md = MessageDigest.getInstance("RIPEMD160");
		md.update(sha256HashByte);
		byte[] ripemd256HashByte = md.digest();
		
		BigInteger ripemdHash = new BigInteger(1, ripemd256HashByte);
		String ripemdHashHex = ripemdHash.toString(16);
		
		while(ripemdHashHex.length()<40){
			
			ripemdHashHex = "0"+ripemdHashHex;
			
		}
		
		System.out.println("ripemd160 applied hash: "+ripemdHashHex+" ("+ripemdHashHex.length()*4+" bits)");
		
		//-------------------------------------------------------
		
		ripemdHashHex = "00"+ripemdHashHex;
		
		System.out.println("adding network bytes to ripemd160 hash - extended ripemd160: "+ripemdHashHex+"\n");
		
		//-------------------------------------------------------
		
		md = MessageDigest.getInstance("SHA-256");
		md.update(hexStringToByte(ripemdHashHex, 21));
		byte[] checking = md.digest();
		
		System.out.println("sha-256 to extended ripemd160: "+new BigInteger(checking).toString(16)+" ("+new BigInteger(checking).toString(16).length()*4+" bits)");
		
		//-------------------------------------------------------
		
		md = MessageDigest.getInstance("SHA-256");
		md.update(checking);
		checking = md.digest();
		
		BigInteger checkingInt = new BigInteger(1, checking); // use this 1 to tell it is positive (https://stackoverflow.com/questions/6357234/sha-hash-function-gives-a-negative-output)
		
		System.out.println("second time sha-256 applied to extended ripe160: "+checkingInt.toString(16)+" ("+checkingInt.toString(16).length()*4+" bits)");
		
		String checksum = checkingInt.toString(16).substring(0,8);
		
		System.out.println("checksum: "+checksum+"\n");
		
		//-------------------------------------------
		
		//add checksum to network bytes added RIPEMD160 hash
		
		String address = ripemdHashHex+checksum;
		
		System.out.println("adding checksum to extended ripemd160 "+address+" ("+address.length()*4+" bits)");
		
		//-------------------------------------------
		
		String base58Address = Base58.encode(new BigInteger(address, 16).toByteArray());
		
		while(base58Address.length() < 34){
			
			base58Address = "1"+base58Address;
			
		}
		
		System.out.println("base 58 bitcoin address: "+base58Address+" ("+base58Address.length()*4+" bits)");
		
	}
	
	public static byte[] hexStringToByte(String hexString, int desiredLength){
		
		byte[] byteTransformation = new BigInteger(hexString, 16).toByteArray();
		
		if(byteTransformation.length < desiredLength){
			
			byte[] paddingApplied = new byte[desiredLength];
			
			//initialize
			for(int i=0;i<desiredLength;i++){
				
				paddingApplied[i] = 0;
				
			}
			
			//transfer
			
			for(int i=0;i<byteTransformation.length;i++){
				
				paddingApplied[i + (desiredLength - byteTransformation.length)] = byteTransformation[i];
				
			}
			
			return paddingApplied;
			
		}
		
		return byteTransformation;
		
	}

}

File: python/bitcoin.py
import ecc
import codecs
import hashlib

#------------------------------------

dump = False
applyBitcoinAddress = True

#------------------------------------
#curve configuration

mod = pow(2, 256) - pow(2, 32) - pow(2, 9) - pow(2, 8) - pow(2, 7) - pow(2, 6) - pow(2, 4) - pow(2, 0)
order = 115792089237316195423570985008687907852837564279074904382605163141518161494337

#curve configuration
# y^2 = x^3 + a*x + b = y^2 = x^3 + 7
a = 0
b = 7

#base point on the curve
x0 = 55066263022277343669578718895168534326250603453777594175500187360389116729240
y0 = 32670510020758816978083085130507043184471273380659243275938904335757337482424

print("---------------------")
print("initial configuration")
print("---------------------")
print("Curve: y^2 = x^3 + ",a,"*x + ",b)
print("Base point: (",x0,", ",y0,")\n")
print("modulo: ", mod)
print("order of group: ", order)

#------------------------------------
#subsdiary functions

def hexStringToByte(content):
	return codecs.decode(content.encode("utf-8"), 'hex')

def hashHex(algorithm, content):
	my_sha = hashlib.new(algorithm)
	my_sha.update(hexStringToByte(content))
	return my_sha.hexdigest()
	
print("-----------------------")
print("generate bitcoin addres")
print("-----------------------")

privateKey = 11253563012059685825953619222107823549092147699031672238385790369351542642469

#print("private key: ", privateKey)
print("private key (hex): ", hex(privateKey)[2:]," (keep this secret!)\n")

publicKey = ecc.applyDoubleAndAddMethod(x0, y0, privateKey, a, b, mod)
#print("public key: ", publicKey)

publicKeyHex = "04"+hex(publicKey[0])[2:]+hex(publicKey[1])[2:]
print("public key (hex): ", publicKeyHex)

publicKeyMerged = int(publicKeyHex, 16)
#print(publicKeyMerged)

print("---------------------------------------")

output = hashHex('sha256', publicKeyHex)	
print("apply sha-256 to public key: ",output)

output = hashHex('ripemd160', output)	
print("apply ripemd160 to sha-256 applied public key: ", output)

output = "00"+output
print("add network bytes to ripemd160 applied hash - extended ripemd160: ", output,"\n")

print("checksum calculation")

checksum = hashHex('sha256', output)
print("apply sha-256 to extended ripemd160: ", checksum)

checksum = hashHex('sha256', checksum)
print("apply second time sha-256: ", checksum)

checksum = checksum[0:8]
print("extract first 8 characters as checksum: ", checksum)

address = output+checksum
print("append checksum to extended ripemd160", address)

print("---------------------------------------")

import base58
address = base58.b58encode(hexStringToByte(address))
print("this is your bitcoin address:",str(address)[2:len(address)-2])
